Client/Frontend/Browser/TabManager.swift

```diff
@@ -232,30 +232,70 @@ class TabManager : NSObject {
 }
 
 extension TabManager {
+    class SavedTab: NSObject, NSCoding {
+        let url: NSURL?
+        let isReaderMode: Bool
+        let isSelected: Bool
+
+        init(url: NSURL, isReaderMode: Bool, isSelected: Bool) {
+            self.url = url
+            self.isReaderMode = isReaderMode
+            self.isSelected = isSelected
+        }
+
+        required init(coder: NSCoder) {
+            self.url = coder.decodeObjectForKey("url") as? NSURL
+            self.isReaderMode = coder.decodeBoolForKey("isReaderMode")
+            self.isSelected = coder.decodeBoolForKey("isSelected")
+        }
+
+        func encodeWithCoder(coder: NSCoder) {
+            coder.encodeObject(url, forKey: "url")
+            coder.encodeBool(isReaderMode, forKey: "isReaderMode")
+            coder.encodeBool(isReaderMode, forKey: "isSelected")
+        }
+    }
+
     func encodeRestorableStateWithCoder(coder: NSCoder) {
-        coder.encodeInteger(count, forKey: "tabCount")
-        coder.encodeInteger(selectedIndex, forKey: "selectedIndex")
-        for i in 0..<count {
-            let tab = tabs[i]
-            coder.encodeObject(tab.url!, forKey: "tab-\(i)-url")
+        var savedTabs = [SavedTab]()
+        for (tabIndex, tab) in enumerate(tabs) {
+            if let url = tab.url, displayURL = tab.displayURL {
```

Comment:
Note: @wesj just landed a change where about:home tabs have a nil `displayURL`, so we'll need to handle this somehow assuming we want to restore about:home tabs.